### PR TITLE
Remove all bump mapping and parallax occlusion related code.

### PR DESCRIPTION
**Pull request by lhofhansl**
See #10479 for discussion.

This PR removes bump mapping, parallax occlusion, and normal mapping code, since it was buggy and not fully implemented.
This simplifies the shader code and hopefully makes it easier to start over.

Note that this also remove the normal texture feature - since it normal textures are no longer used for anything.

### CODE REVIEW

**Comment by paramat** [comment]
I can review and test this, although i feel a little underqualified in some areas of the changes =)

**Comment by lhofhansl** [comment]
@HybridDog FYI since you've worked in this area.

File: builtin/settingtypes.txt [hunk 1]
```diff
@@ -711,9 +686,6 @@ fall_bobbing_amount (Fall bobbing factor) float 0.03
 #    Note that the interlaced mode requires shaders to be enabled.
 3d_mode (3D mode) enum none none,anaglyph,interlaced,topbottom,sidebyside,crossview,pageflip
 
-#    Strength of 3D mode parallax.
```
Comment (paramat) [inline builtin/settingtypes.txt:714]: This is actually unrelated to shaders, it is for 3D rendering modes.

Comment (lhofhansl) [reply builtin/settingtypes.txt:714]: Whoops. Yes. I'll bring that back.

**Comment by paramat** [comment]
Interesting how the documentation reveals these shaders are actually 'experimental', unfinished and buggy. It is more of a mess than i realised.
> Bumpmapping and parallax occlusion are currently experimental features:
* Parallax occlusion with relief-mapping mode does not yet work correctly
  together with Minetest's Fastfaces.

Oddly, relief mapping mode is actually set as the default mode, and fast faces are everywhere.

File: builtin/settingtypes.txt (hunk 1 above)
Comment (lhofhansl) [reply builtin/settingtypes.txt]: Done

File: src/client/mesh_generator_thread.cpp [hunk 2]
```diff
@@ -207,8 +204,7 @@ CachedMapBlockData* MeshUpdateQueue::getCachedBlock(const v3s16 &p)
 
 void MeshUpdateQueue::fillDataFromMapBlockCache(QueuedMeshUpdate *q)
 {
-	MeshMakeData *data = new MeshMakeData(m_client, m_cache_enable_shaders,
-			m_cache_use_tangent_vertices);
+	MeshMakeData *data = new MeshMakeData(m_client, m_cache_enable_shaders);
```
Comment (paramat) [inline src/client/mesh_generator_thread.cpp:207]: Perhaps the 3rd argument should be removed from `MeshMakeData()` https://github.com/minetest/minetest/search?q=MeshMakeData
... plus the related changes, such as removing `m_use_tangent_vertices` https://github.com/minetest/minetest/search?q=m_use_tangent_vertices
and `m_cache_use_tangent_vertices` https://github.com/minetest/minetest/search?q=m_cache_use_tangent_vertices

Comment (lhofhansl) [reply src/client/mesh_generator_thread.cpp:207]: I was debating that internally. Then decided not to, but you're right... I'll remove that too.

Comment (lhofhansl) [reply src/client/mesh_generator_thread.cpp:207]: Done. Removed even more code.

**Comment by paramat** [comment]
I will make this my top priority for review.

**Comment by lhofhansl** [comment]
@paramat Thanks! You already caught two issues - I've been sloppy.

**Comment by lhofhansl** [comment]
@TheTermos does not feel that we should remove the parallax occlusion code.

**Comment by TheTermos** [comment]
> @TheTermos does not feel that we should remove the parallax occlusion code.

Where did that come from?

**Comment by lhofhansl** [comment]
@TheTermos Oops. That was @HybridDog. I truly apologize!

**Comment by paramat** [comment]
My testing in the issue which shows pixel wrapping reminded me of something i thought i would mention.

The wrapping of pixels from the opposite edge, issue #2815 , was fixed in https://github.com/minetest/minetest/commit/655fc6010ffd4be7de315be261df2a61d5d4538a by adding the tiledef fields 'tileable_vertical', 'tileable_horizontal'.

Perhaps we should leave these tiledef fields present in code, for now at least, as they are likely to be needed by any new implementation of bumpmapping / parallax occlusion. However perhaps we should remove mention of them in lua_api.txt as they will be irrelevant?

We might decide to remove these tiledef fields, but i suggest that we consider and do that in a later PR, to not bloat and delay this PR.

**Comment by lhofhansl** [comment]
Thanks @paramat, I don't see that I'm removing any tiledef fields in this PR. Happy to shrink the scope of this, which part are you referring to?

**Comment by paramat** [comment]
> I don't see that I'm removing any tiledef fields in this PR

Sorry to be unclear, i do not mean to imply you are removing those in this PR, you are not. I am just mentioning the related tilespec code and suggesting we leave that in, for now at least.
So the only change to the PR i am suggesting is removing the documentation of 'tileable_vertical/horizontal' from lua_api.txt.

File: client/shaders/object_shader/opengl_fragment.glsl [hunk 3]
```diff
@@ -89,13 +89,6 @@ void main(void)
 	bool use_normalmap = false;
 	get_texture_flags();
```
Comment (paramat) [inline client/shaders/object_shader/opengl_fragment.glsl:90]: Remove?

**Comment by paramat** [comment]
In client/shaders/object_shader/opengl_fragment.glsl, do the following need removing?
```
bool normalTexturePresent = false;
bool texTileableHorizontal = false;
bool texTileableVertical = false;
bool texSeamless = false;
```
```
void get_texture_flags()
{
	vec4 flags = texture2D(textureFlags, vec2(0.0, 0.0));
	if (flags.r > 0.5) {
		normalTexturePresent = true;
	}
	if (flags.g > 0.5) {
		texTileableHorizontal = true;
	}
	if (flags.b > 0.5) {
		texTileableVertical = true;
	}
	if (texTileableHorizontal && texTileableVertical) {
		texSeamless = true;
	}
}
vec4 get_normal_map(vec2 uv)
{
	vec4 bump = texture2D(normalTexture, uv).rgba;
	bump.xyz = normalize(bump.xyz * 2.0 - 1.0);
	return bump;
}
```

**Comment by paramat** [comment]
In  client/shaders/nodes_shader/opengl_fragment.glsl, do the following need removing?
```
uniform sampler2D normalTexture;
uniform sampler2D textureFlags;
```

**Comment by paramat** [comment]
In  client/shaders/nodes_shader/opengl_vertex.glsl, do the following need removing?
```
uniform vec3 eyePosition;
```

**Comment by paramat** [comment]
I tested. Main menu is fine and everything seems to work ok.
Reviewed code and all looks fine, apart from the extra removals needed from shader files detailed above.
Will retest after those are attended to and hopefully then approve.

**Comment by lhofhansl** [comment]
Arrgghh. Thanks @paramat . I'll do a second pass through what can be removed and I'll remove the tilespec doc.

I wasn't sure if I wanted to remove all the uniforms from the shaders. I'll double check these.

**Comment by paramat** [comment]
> I wasn't sure if I wanted to remove all the uniforms from the shaders

Ok i see. Although it will all still be in history to refer too, so not sure if there is advantage to leaving some stuff in creating a confusing mess.

I remember that 'vertex tangent' stuff was added by RBA purely for bumpmapping and parallax occlusion, in this commit https://github.com/minetest/minetest/commit/43fcfbfe05578d7471d40c8c087fd04e24b264b5

I did a search for 'tangents' in current code an found some extra stuff not in your PR:
https://github.com/minetest/minetest/blob/9bba52c4000a06043f5100dbb0ef66d869707ffc/src/client/mesh.cpp#L381
https://github.com/minetest/minetest/blob/9bba52c4000a06043f5100dbb0ef66d869707ffc/src/client/mesh.cpp#L978

**Comment by lhofhansl** [comment]
I feel the following needs to be left in:
- case video::EVT_TANGENTS: since this is a valid enum member 
- vNormal, since this is generally useful for the fragment shader to have.
- eyePosition, as this is also generally good to have

**Comment by lhofhansl** [comment]
The failure on Android appears unrelated.

**Comment by paramat** [comment]
> I feel the following needs to be left in:

That seems fine to me. I think this has a good balance of how much to remove.
Also i think we should not get too perfectionist for this first PR, as long as everything works, more could be cleaned up in a follow up PR.
I will retest, rereview and hopefully approve as soon as possible.

File: client/shaders/object_shader/opengl_fragment.glsl (hunk 3 above)
Comment (paramat) [reply client/shaders/object_shader/opengl_fragment.glsl]: Is done.

**Comment by paramat** [comment]
A related warning during compilation:
```
/src/client/game.cpp: In member function ‘virtual void GameGlobalShaderConstantSetter::onSetConstants(irr::video::IMaterialRendererServices*, bool)’:
/src/client/game.cpp:553:5: warning: unused variable ‘normal_tex’ [-Wunused-variable]
     normal_tex = 1,
     ^~~~~~~~~~
/src/client/game.cpp:554:5: warning: unused variable ‘flags_tex’ [-Wunused-variable]
     flags_tex = 2;
     ^~~~~~~~~
```

**Comment by paramat** [comment]
All my previous comments are attended to, rereviewed code, retested, all seems fine.
Once you fix the warnings and retest you can assume my approval and merge.

**Comment by lhofhansl** [comment]
I saw that warning too, but I did not want remove that enum. Hmm... I could.
Also, since the Android failure is persistent and I do not see this with other PRs perhaps it is pertinent?

**Comment by paramat** [comment]
> FAILURE: Build failed with an exception.
Task :native:downloadSqlite FAILED
Download https://www.sqlite.org/2020/sqlite-amalgamation-3320200.zip
* What went wrong:
Execution failed for task ':native:downloadSqlite'.
Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
> de.undercouch.gradle.tasks.download.org.apache.http.conn.HttpHostConnectException: Connect to www.sqlite.org:443 [www.sqlite.org/2600:3c00:0:0:f03c:91ff:fe96:b959] failed: Network is unreachable (connect failed)

It looks like a connection failure unrelated to these changes, so being persistent and this PR only seems odd. Maybe @rubenwardy could advise?

**Comment by rubenwardy** [comment]
That's a temporary error due to website downtime, I've restarted the build

**Comment by paramat** [comment]
Good to know, and thanks.
So i think this should be merged if builds are successful, before conflicts start arising and so i do not have to rerereview. We can always do some additional clean up in a follow up PR.

**Comment by paramat** [comment]
Can be merged now.

**Comment by lhofhansl** [comment]
Actually lemme look at that enum. If it can be safely removed I'll do that and then merge. Otherwise I'll merge as is.

**Comment by lhofhansl** [comment]
It's not an enum anyway, I just removed he unused variables, which avoids that warning, going to merge soon.

**Comment by paramat** [comment]
Ok looks fine.
